### PR TITLE
electron-205: fixed file overwrite issues when a custom download folder is set by the user

### PR DESCRIPTION
**Pull request by VishwasShashidhar**
## Description
Supports functionality to set a custom downloads folder in electron [electron-205](https://perzoinc.atlassian.net/browse/electron-205)

Although this was already done, there was an issue where in with a custom downloads folder, an existing file used to be overwritten unlike chrome which appends a number to the new filename in case the file already exists.

## Approach
How does this change address the problem?
- #### Problem with the code: Electron's default download mechanism doesn't support the increment behaviour for a custom folder
- #### Fix : Wrote a custom logic to support the above functionality

## Learning
N/A

#### Blog Posts
N/A

## Related PRs
N/A

## Open Questions if any and Todos
N/A

### CODE REVIEW

**Comment by VishwasShashidhar** [comment]
@VikasShashidhar Please review.

CC: @KiranNiranjan

File: js/windowMgr.js [hunk 1]
```diff
@@ -262,21 +257,29 @@ function doCreateMainWindow(initialUrl, initialBounds) {
...
+        if (downloadsDirectory !== defaultDownloadsDirectory && !fs.existsSync(downloadsDirectory)) {
+            downloadsDirectory = defaultDownloadsDirectory;
+            updateConfigField("downloadsDirectory", downloadsDirectory);
         }
         
-        // Send file path when download is complete
+        // We check the downloads directory to see if a file with the similar name
+        // already exists and get a unique filename if that's the case
+        let newFileName;
+        newFileName = getUniqueFileName(item.getFilename());
```
Comment (VikasShashidhar) [inline js/windowMgr.js]: The above 2 lines can be combined.

File: js/windowMgr.js [hunk 2]
```diff
@@ -756,6 +759,53 @@ function repositionMainWindow() {
...
+ * @param filename filename passed by the remote server
+ * @returns {String} the new filename
+ */
+function getUniqueFileName(filename) {
+    
+    // By default, we assume that the file exists
+    const fileExists = true;
+    
+    // We break the file from it's extension to get the name
+    const fileName = filename.substr(0, filename.lastIndexOf('.')) || filename;
```
Comment (VikasShashidhar) [inline js/windowMgr.js]: Please use a different name. Using "fileName" and "filename" is confusing.

File: js/windowMgr.js [hunk 3]
```diff
@@ -756,6 +759,53 @@ function repositionMainWindow() {
...
+    const fileType = filename.split('.').pop();
+    
+    // We use this to set the new file name with an increment on the previous existing file
+    let fileNumber = 0;
+    let newPath;
+    
+    while (fileExists) {
+        
+        let fileNumber_str = fileNumber.toString();
```
Comment (VikasShashidhar) [inline js/windowMgr.js]: Remove "_" from name.

File: js/windowMgr.js [hunk 4]
```diff
@@ -756,6 +759,53 @@ function repositionMainWindow() {
...
+        
+        // By default, we know if the file doesn't exist,
+        // we can use the filename sent by the remote server
+        let current = filename;
+        
+        // If the file already exists, we know that the
+        // file number variable is increased, so,
+        // we construct a new file name with the file number
+        if (fileNumber > 0) {
+            current = fileName + " (" + fileNumber_str + ")." + fileType;
```
Comment (VikasShashidhar) [inline js/windowMgr.js]: Could use ES6 template literals to simplify this.

**Pull request merged**